src/Storages/ExternalStream/Kafka/WriteBufferFromKafka.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <IO/BufferWithOwnMemory.h>
+#include <IO/WriteBuffer.h>
+#include <KafkaLog/KafkaWALCommon.h>
+
+namespace DB
+{
+
+class WriteBufferFromKafka final : public BufferWithOwnMemory<WriteBuffer>
+{
+public:
+    WriteBufferFromKafka(klog::KTopicPtr topic_, Poco::Logger * logger, size_t buffer_size = DBMS_DEFAULT_BUFFER_SIZE);
+
+    ~WriteBufferFromKafka() override = default;
+
+    void onMessageDelivery(const rd_kafka_message_t *);
+
+    /// the number of acknowledgement has been received so far for the current checkpoint period
+    ALWAYS_INLINE size_t acked() const { return state.acked; }
```

Comment:
When we are in head files, not need to specify `ALWAYS_INLINE`

Reply:
and these routines are not called in critical loop path

Reply:
Good to know.